lib/api/funnel.js

```diff
@@ -241,64 +246,65 @@ class Funnel {
       return 1;
     }
 
-    const processNow = this.getRequestSlot('execute', request, callback);
+    try {
+      const executing = this.throttle(req => {
+        if (req.error) {
+          // "handleErrorDump" shouldn't need to be called for 503 errors
+          callback(req.error, req);
+          return;
+        }
 
-    if (request.error) {
-      // "handleErrorDump" shouldn't need to be called for 503 errors
-      callback(request.error, request);
-      return 1;
-    }
+        // if the connection is closed there is no need to execute the request
+        // => discarding it
+        if (!this.kuzzle.router.isConnectionAlive(req.context)) {
+          debug('Client connection dead: dropping request: %a', req.input);
+          callback(processError.get('connection_dropped'), req);
+          return;
+        }
 
-    // request has been cached. Do not process now
-    if (!processNow) {
-      return -1;
-    }
+        debug('Starting request %s:%s [%s]: %j', req.input.controller, req.input.action, req.id, req.input);
+
+        let _request;
+
+        this.kuzzle.asyncStore.run(() => {
+          this.kuzzle.asyncStore.set('REQUEST', req);
+
+          this.checkRights(req)
+            .then(modifiedRequest => {
+              _request = modifiedRequest;
+              return this.rateLimiter.isAllowed(_request);
+            })
+            .then(allowed => {
+              if (!allowed) {
+                throw processError.get('too_many_requests');
+              }
+
+              return this.processRequest(_request);
+            })
+            .then(processResult => {
+              debug('Request %s successfully executed. Result: %a',
+                req.id,
+                processResult);
+
+              callback(null, processResult);
+
+              // disables a bluebird warning in dev. mode triggered when
+              // a promise is created and not returned
+              return null;
+            })
+            .catch(err => {
+              debug('Error processing request %s: %a', req.id, err);
+              return this._executeError(err, req, true, callback);
+            });
+        });
+      }, this, request);
 
-    // if the connection is closed there is no need to execute the request
-    // => discarding it
-    if (!this.kuzzle.router.isConnectionAlive(request.context)) {
-      debug('Client connection dead: dropping request: %a', request.input);
-      callback(processError.get('connection_dropped'), request);
+      return executing ? 0 : -1;
+    } catch (error) {
+      request.setError(error);
```

Comment:
`throttle` only throws when refusing a request with a 503 error. When it does, it already sets the right error object in the request, so doing so here is redundant. 
You have to either remove that line, or the `request.setError` one performed by `throttle`.

Reply:
This point tinkers me a bit (which is why I left this redundant behavior).
* On one side, I think `throttle` should be responsible of erroring a request if it is discarded,
* On the other side, I find it a bit obscure to let `throttle` mutate an object passed as argument: how do you make it explicit? 

In other words, throwing or returning is the most explicit we can do, but we'll have to remember to manually `request.setError` upon rejection. Mutating the request inside `throttle` does the job for us but we'll have to remember this implicit "side-effect".

What do you think?

Reply:
For the moment, I've choosen the explicit option.

Reply:
I like explicit behaviors.